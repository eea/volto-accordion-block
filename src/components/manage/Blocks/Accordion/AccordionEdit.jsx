import cx from 'classnames';
import React from 'react';
import AnimateHeight from 'react-animate-height';
import { Accordion, Input } from 'semantic-ui-react';
import { Icon } from './util';
import config from '@plone/volto/registry';
import { defineMessages, injectIntl } from 'react-intl';

const messages = defineMessages({
  EnterTitle: {
    id: 'Enter Title',
    defaultMessage: 'Enter Title',
  },
});

const AccordionEdit = (props) => {
  const {
    children,
    handleTitleChange,
    handleTitleClick,
    uid,
    panel,
    data,
    index,
    intl,
  } = props;
  const [activeIndex, setActiveIndex] = React.useState([0]);
  const accordionConfig = config.blocks.blocksConfig.accordion;
  const { titleIcons } = accordionConfig;
  const isActive = activeIndex.includes(index);
  const iconOnRight = data.right_arrows;
  const iconPosition = iconOnRight ? 'rightPosition' : 'leftPosition';

  const handleClick = (e, itemProps) => {
    const { index } = itemProps;
    if (data.non_exclusive) {
      const newIndex =
        activeIndex.indexOf(index) === -1
          ? [...activeIndex, index]
          : activeIndex.filter((item) => item !== index);

      setActiveIndex(newIndex);
    } else {
      const newIndex =
        activeIndex.indexOf(index) === -1
          ? [index]
          : activeIndex.filter((item) => item !== index);

      setActiveIndex(newIndex);
    }
  };

  React.useEffect(() => {
    return data.collapsed ? setActiveIndex([]) : setActiveIndex([0]);
  }, [data.collapsed]);

  return (
    <Accordion
      className={
        data.styles ? data.styles.theme : accordionConfig?.defaults?.theme
      }
      {...accordionConfig.options}
    >
      <React.Fragment>
        <Accordion.Title
          as={data.title_size}
          active={isActive}
          index={index}
          onClick={handleClick}
          className={cx('accordion-title', {
            'align-arrow-left': !iconOnRight,
            'align-arrow-right': iconOnRight,
          })}
        >
          <Icon
            options={titleIcons}
            name={
              isActive
                ? titleIcons.opened[iconPosition]
                : titleIcons.closed[iconPosition]
            }
          />
          {!data.readOnlyTitles ? (
            <Input
              fluid
              className="input-accordion-title"
              transparent
              placeholder={intl.formatMessage(messages.EnterTitle)}
              value={panel?.title || ''}
              onClick={(e) => {
                handleTitleClick();
                e.stopPropagation();
              }}
              onChange={(e) => handleTitleChange(e, [uid, panel])}
            />
          ) : (
            <span>{panel?.title}</span>
          )}
        </Accordion.Title>
        <AnimateHeight
          animateOpacity
          duration={500}
          height={isActive ? 'auto' : 0}
        >
          <Accordion.Content active={isActive}>{children}</Accordion.Content>
        </AnimateHeight>
      </React.Fragment>
    </Accordion>
  );
};

export default injectIntl(AccordionEdit);
