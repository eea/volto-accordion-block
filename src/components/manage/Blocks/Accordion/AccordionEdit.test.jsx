import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import AccordionEdit from './AccordionEdit';
import renderer from 'react-test-renderer';
import config from '@plone/volto/registry';
import { Provider } from 'react-intl-redux';
import configureStore from 'redux-mock-store';
import '@testing-library/jest-dom/extend-expect';

const mockStore = configureStore();

config.blocks.blocksConfig.accordion = {
  ...config.blocks.blocksConfig.accordion,
  defaults: {
    theme: 'defaultTheme',
  },
  titleIcons: {
    opened: {
      rightPosition: 'chevron left',
      leftPosition: 'chevron right',
    },
    closed: {
      rightPosition: 'chevron down',
      leftPosition: 'chevron down',
    },
    size: 'tiny',
    iconComponent: 'SemanticIcon',
  },
};

const handleTitleChange = jest.fn();
const handleTitleClick = jest.fn();

const store = mockStore({
  intl: {
    locale: 'en',
    messages: {},
    formatMessage: () => 'Select layout',
  },
});
describe('AccordionEdit', () => {
  const uid = 'uid';
  const panel = { title: 'Panel Title' };
  const data = {
    title_size: 'h3',
    right_arrows: true,
    collapsed: false,
    exclusive: false,
    non_exclusive: true,
    styles: {
      theme: 'default',
    },
  };
  const data1 = {
    title_size: 'h3',
    right_arrows: false,
    collapsed: true,
    exclusive: false,
    non_exclusive: true,
    readOnlyTitles: true,
  };
  const index = 0;

  it('should render correctly', () => {
    const component = renderer.create(
      <Provider store={store}>
        <AccordionEdit
          handleTitleChange={handleTitleChange}
          handleTitleClick={handleTitleClick}
          uid={uid}
          panel={panel}
          data={data}
          index={index}
        />
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render correctly', () => {
    const { getByText } = render(
      <Provider store={store}>
        <AccordionEdit
          handleTitleChange={handleTitleChange}
          handleTitleClick={handleTitleClick}
          uid={uid}
          panel={panel}
          data={data1}
          index={index}
        />
      </Provider>,
    );

    expect(getByText(panel.title)).toBeInTheDocument();
  });

  it('should handle title click correctly', () => {
    const { getByDisplayValue } = render(
      <Provider store={store}>
        <AccordionEdit
          handleTitleChange={handleTitleChange}
          handleTitleClick={handleTitleClick}
          uid={uid}
          panel={panel}
          data={data}
          index={index}
        />
      </Provider>,
    );

    fireEvent.click(getByDisplayValue(panel.title));

    expect(handleTitleClick).toHaveBeenCalled();
  });

  it('should handle title change correctly and check the parameters the handleTitleChange function was called with', () => {
    const { getByDisplayValue } = render(
      <Provider store={store}>
        <AccordionEdit
          handleTitleChange={handleTitleChange}
          handleTitleClick={handleTitleClick}
          uid={uid}
          panel={panel}
          data={data}
          index={index}
        />
      </Provider>,
    );

    fireEvent.change(getByDisplayValue(panel.title), {
      target: { value: 'New Panel Title' },
    });

    expect(handleTitleChange).toHaveBeenCalledWith(expect.any(Object), [
      uid,
      panel,
    ]);
  });

  it('should open accordion content when title is clicked', () => {
    config.blocks.blocksConfig.accordion = {
      ...config.blocks.blocksConfig.accordion,
    };
    const { container, getByText } = render(
      <Provider store={store}>
        <AccordionEdit
          handleTitleChange={handleTitleChange}
          handleTitleClick={handleTitleClick}
          uid={uid}
          panel={panel}
          data={data}
          index={index}
        >
          <p>Accordion Content</p>
        </AccordionEdit>
      </Provider>,
    );
    const accordionTitle = container.querySelector('.accordion-title');
    fireEvent.click(accordionTitle);

    const contentElement = getByText('Accordion Content');
    expect(contentElement).toBeInTheDocument();
  });

  it('should open accordion content when title is clicked', () => {
    const { container, getByText } = render(
      <Provider store={store}>
        <AccordionEdit
          handleTitleChange={handleTitleChange}
          handleTitleClick={handleTitleClick}
          uid={uid}
          panel={panel}
          data={{ ...data, non_exclusive: false }}
          index={index}
        >
          <p>Accordion Content</p>
        </AccordionEdit>
      </Provider>,
    );
    const accordionTitle = container.querySelector('.accordion-title');
    fireEvent.click(accordionTitle);

    const contentElement = getByText('Accordion Content');
    expect(contentElement).toBeInTheDocument();
  });
});
