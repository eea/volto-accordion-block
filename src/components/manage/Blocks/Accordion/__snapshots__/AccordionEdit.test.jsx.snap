// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionEdit should render correctly 1`] = `
<div
  className="accordion ui default"
>
  <h3
    className="active title accordion-title align-arrow-right"
    onClick={[Function]}
  >
    <i
      aria-hidden="true"
      className="chevron left tiny icon"
      onClick={[Function]}
    />
    <div
      className="ui fluid transparent input input-accordion-title"
    >
      <input
        onChange={[Function]}
        onClick={[Function]}
        placeholder="Enter Title"
        type="text"
        value="Panel Title"
      />
    </div>
  </h3>
  <div
    aria-hidden={false}
    className="rah-static rah-static--height-auto"
    style={
      Object {
        "height": "auto",
        "overflow": "visible",
      }
    }
  >
    <div
      style={
        Object {
          "WebkitTransition": "opacity 500ms ease 0ms",
          "transition": "opacity 500ms ease 0ms",
        }
      }
    >
      <div
        className="content active"
      />
    </div>
  </div>
</div>
`;
